### PR TITLE
Protocol fee adjustment in driver

### PR DESCRIPTION
**Pull request by sunce86**
# Description
Replacement PR for https://github.com/cowprotocol/services/pull/2097

As @fleupold suggested, protocol fee adjustments are done fully in `driver`, in class `Fulfillment`. It goes like this:
1. Solver responds with solution (accounts for surplus_fee only)
2. Driver creates the fulfillment (executed amount + surplus_fee) based on (1)
3. Driver adjusts fulfillment for protocol fees. It
- calculates protocol fee
- reduces executed amount by protocol fee.
(note that this is basically the same as increasing `surplus_fee`)
4. Driver encodes the fulfillment using the `SettlementEncoder` functionality, meaning custom price adjustments will be reused for protocol fee also.

## How to test
Added unit tests for
1. sell limit order fok
2. buy limit order fok
3. sell limit order partial (with 3 fulfillments)
(but now removed to keep the domain clean)

### CODE REVIEW

File: crates/driver/src/domain/competition/solution/trade.rs
Comment (fleupold) [inline crates/driver/src/domain/competition/solution/trade.rs:83]: Have we decided/documented if this factor is supposed to be interpreted as the % of surplus of the "raw solution" (which we only have access to if solvers use our driver) or the % of remaining surplus after the fee is applied. I'd prefer the latter (since this is what we will observe on chain) but here it seems to be the former.

Let me explain with an example (of a driver not using our logic). The protocol fee is 50%. A user wants to sell at most 120 USDC for 100 DAI. On chain we observe 110USDC for 100 DAI. My assumption is that we now charge the solver 5USDC (50% of the observed surplus).

This suggests the true price the solver found (before any fee) must have been something like: sell 105 USDC for 100 DAI

However, if they now used our driver logic with this route, we would take 50% of the 15USDC raw solution surplus and thus withhold 7.5USDC instead of the desired 5.

File: crates/driver/src/domain/competition/solution/trade.rs
Comment (fleupold) [inline crates/driver/src/domain/competition/solution/trade.rs:119]: I'd factor the logic to get "surplus_in_sell_token" into a separate method and then combine the price_improvement_fee computation and min for both match arms.

File: crates/driver/src/domain/competition/solution/trade.rs
Comment (fleupold) [inline crates/driver/src/domain/competition/solution/trade.rs:127]: We should be able to reuse this logic for volume based fee policies, right? Can we have `volume_fee` method that can be reused across arms?

**Review by fleupold — CHANGES_REQUESTED** [review]
I don't feel comfortable adding this large chunk of core logic right into the existing fulfilment constructor. 
Can we create a fee specific domain module (domain/competition/solution/fee.rs) and add an extension `impl Fulfilment` there which exposes an `apply_fee(self, ...) -> Self` that gets triggered from the solution (either upon constructing or before being encoded)?
This module can then encapsulate all fee related logic.

As for testing, I think one nice side effect of the examples you worked on with Olga and Anxo is that they can serve as ideal e2e tests, or at least driver integration tests, which is how we want to test core functionality going forward (instead of unit tests).

File: crates/driver/src/domain/competition/solution/trade.rs
Comment (sunce86) [reply crates/driver/src/domain/competition/solution/trade.rs:83]: In this equation the `surplus` is calculated as raw surplus - surplus fee, so it's what you prefer. It's because we use executed_sell_amount_**with_surplus_fee** for `surplus` calculation.

**Comment by sunce86** [comment]
> I don't feel comfortable adding this large chunk of core logic right into the existing fulfilment constructor.
> Can we create a fee specific domain module (domain/competition/solution/fee.rs) and add an extension impl Fulfilment there which exposes an apply_fee(self, ...) -> Self that gets triggered from the solution (either upon constructing or before being encoded)?
> This module can then encapsulate all fee related logic.

Tried this already but disliked the fact that the `Fulfillment` object has two states and I would have to make field `Fulfillment::protocol_fee` optional. Since you prefer to not even have this field, then this is doable.

File: crates/driver/src/domain/competition/solution/trade.rs
Comment (sunce86) [reply crates/driver/src/domain/competition/solution/trade.rs:26]: No other reason. I liked everything in one place because if I move out protocol fee calculation, then `Fullfilment` object will have two states, one when receives raw solution from solver, other after protocol fee is applied (wherever that happens) so one needs to be aware of this when debugging and maintaining code.
But in general I am ok with both.

File: crates/driver/src/domain/competition/solution/trade.rs
Comment (fleupold) [reply crates/driver/src/domain/competition/solution/trade.rs:83]: Then I'm missing something. We can assume the example is on Gnosis Chain (and ignore any network or otherwise pre-specified surplus fee).

How would a route that sells 105 USDC for 100 DAI be adjusted? It would add 7.5 USDC of surplus fee, which would lead to a net settlement of 112.5USDC for 100 DAI. If observed in the autopilot, this would cause a 50% of 7.5USDC = 3.75USDC bill to be sent to the solver. Thus we overcharged the fee.

Looking at the [e2e example](https://www.notion.so/cownation/Fee-Implementation-End-to-end-f8cfc251c63c416aa07aaca1cebdad1b) it does seem that version 1 was agreed (50% protocol fee means we keep 50% of the surplus before fees). However, the example is lacking detail on how the protocol fee is later on recovered from on chain (where a de-facto 50% protocol fee would translate into taking 100% of the observed surplus).

File: crates/driver/src/domain/competition/solution/trade.rs
Comment (sunce86) [reply crates/driver/src/domain/competition/solution/trade.rs]: resolved.

File: crates/driver/src/domain/competition/order/mod.rs
Comment (sunce86) [reply crates/driver/src/domain/competition/order/mod.rs]: deleted.

File: crates/driver/src/domain/competition/solution/trade.rs
Comment (sunce86) [reply crates/driver/src/domain/competition/solution/trade.rs]: refactored.

File: crates/driver/src/domain/competition/solution/trade.rs
Comment (sunce86) [reply crates/driver/src/domain/competition/solution/trade.rs]: deleted.

File: crates/driver/src/infra/solver/dto/solution.rs
Comment (sunce86) [reply crates/driver/src/infra/solver/dto/solution.rs]: refactored.

File: crates/driver/src/domain/competition/solution/trade.rs
Comment (sunce86) [reply crates/driver/src/domain/competition/solution/trade.rs]: refactored but not sure if still valid comment?

File: crates/driver/src/domain/competition/solution/trade.rs
Comment (sunce86) [reply crates/driver/src/domain/competition/solution/trade.rs]: refactored.

**Comment by sunce86** [comment]
Refactored, requesting review again.

File: crates/driver/src/domain/competition/solution/fee.rs
Comment (fleupold) [inline crates/driver/src/domain/competition/solution/fee.rs:26]: nanonit: zero is more explicitly what we want than default
```suggestion
            Fee::Static => eth::U256::zero(),
```

File: crates/driver/src/domain/competition/solution/fee.rs
Comment (fleupold) [inline crates/driver/src/domain/competition/solution/fee.rs:125]: Can we make these methods operate on self (to avoid passing 5 extra arguments) and have one method for both buy/sell to further simplify the remaining very large match statement in the method above?

File: crates/driver/src/domain/competition/solution/mod.rs
Comment (fleupold) [inline crates/driver/src/domain/competition/solution/mod.rs:195]: I'm pretty sure we shouldn't fail silently here. We might not even want to continue with this solution as it could potentially cause significant damage if a solver doesn't withhold the correct amount of fee.

File: crates/driver/src/domain/competition/solution/fee.rs
Comment (fleupold) [inline crates/driver/src/domain/competition/solution/fee.rs:21]: nit: a small helper `ClearingPrice` helper type with fields buy and sell would make this nicer to read.

File: crates/driver/src/domain/competition/solution/trade.rs
Comment (fleupold) [inline crates/driver/src/domain/competition/solution/trade.rs:98]: I think a comment is needed here, but in its current form it doesn't really help me much given the function name. What do we mean by raw fee? Is this another type of fee we need to define in our glossary? How is it different to just `fee` ?

File: crates/driver/src/domain/competition/solution/fee.rs
Comment (fleupold) [inline crates/driver/src/domain/competition/solution/fee.rs:18]:
```suggestion
    pub fn with_protocol_fee(
```

File: crates/driver/src/domain/competition/solution/fee.rs
Comment (fleupold) [inline crates/driver/src/domain/competition/solution/fee.rs:95]: If we use checked_mul do we also need to use checked_add ? I can see how this can become annoying when using iter functions instead of good old loops. Feel free to change back to whatever is easier (as long as the loop body is not overly complex)

File: crates/driver/src/domain/competition/solution/fee.rs
Comment (fleupold) [inline crates/driver/src/domain/competition/solution/fee.rs:34]: nit: move this into a separate method (on self) to make the main `with_protocol_function` easy to read?

**Review by fleupold — COMMENTED** [review]
Thanks, I like this approach much more. I still think we can factor the code inside fee.rs to make it easier to read and follow but it's mainly style preference.

I don't think the error handling is correct yet, and overflow handling for additions may be missing.

File: crates/driver/src/domain/competition/solution/mod.rs
Comment (sunce86) [reply crates/driver/src/domain/competition/solution/mod.rs:195]: Yes, was not sure about this, will add error handling.

File: crates/driver/src/domain/competition/solution/fee.rs
Comment (sunce86) [reply crates/driver/src/domain/competition/solution/fee.rs]: True. fixed

File: crates/driver/src/domain/competition/solution/fee.rs
Comment (sunce86) [reply crates/driver/src/domain/competition/solution/fee.rs]: Refactored.

File: crates/driver/src/domain/competition/solution/mod.rs
Comment (sunce86) [reply crates/driver/src/domain/competition/solution/mod.rs]: Added error handling

File: crates/driver/src/domain/competition/solution/fee.rs
Comment (sunce86) [reply crates/driver/src/domain/competition/solution/fee.rs]: Added.

File: crates/driver/src/domain/competition/solution/trade.rs
Comment (sunce86) [reply crates/driver/src/domain/competition/solution/trade.rs]: Added a different comment, hopefully better.

File: crates/driver/src/domain/competition/solution/fee.rs
Comment (sunce86) [reply crates/driver/src/domain/competition/solution/fee.rs]: added checked function,

File: crates/driver/src/domain/competition/solution/fee.rs
Comment (sunce86) [reply crates/driver/src/domain/competition/solution/fee.rs]: refactored

File: crates/driver/src/domain/competition/solution/trade.rs
Comment (sunce86) [reply crates/driver/src/domain/competition/solution/trade.rs]: We resolved this right? We take 50% of the surplus after fee (surplus_fee)

File: crates/driver/src/domain/competition/solution/trade.rs
Comment (sunce86) [reply crates/driver/src/domain/competition/solution/trade.rs]: refactored completely so closing

File: crates/driver/src/domain/competition/solution/fee.rs
Comment (fleupold) [inline crates/driver/src/domain/competition/solution/fee.rs:66]: Probably not an issue for now, but I wonder if the current logic would work for multiple fee policies (e.g. two 50% price improvements). I would think those would be applied sequentially (ie. we first get a fulfillment from applying the first fee - with less surplus! - and then only take 50% of the remaining surplus rather than 100%).

My suggestion would be to expose another `apply` method on a local impl extension of the FeePolicy enum, that takes a fulfilment and prices and returns a new fulfilment with its policy applied.

But we can also cross that bridge when we get to it, in this case we should probably document and guard with an error that multiple fee policies aren't yet properly implemented.

**Review by fleupold — COMMENTED** [review]
Code looks good to me (just two small issues with static orders and multiple fee policies).

How do you want to go about adding tests? Are we going to keep the unit tests? In any case, I'd like to see a driver or integration test with the two examples we have in notion to make sure this works without coupling the test to the concrete code factoring.

File: crates/driver/src/domain/competition/solution/fee.rs
Comment (sunce86) [reply crates/driver/src/domain/competition/solution/fee.rs:145]: Solvers could submit solutions with violated limit prices. I just didn't want this part of the code to detect this kind of errors. I want this module to be responsible for 1 thing - calculating protocol fees, and let simulation verify the solution like it did before.

File: crates/driver/src/domain/competition/solution/fee.rs
Comment (sunce86) [reply crates/driver/src/domain/competition/solution/fee.rs:66]: I need to think more if we should allow for multiple policies of the same kind. Does it make sense to have two price improvement policies on the same order, where the calculated fee amounts depend on the ordering of policies.
Would you expect two policies, one price improvement, one volume based, to also be sequentially applied?

File: crates/driver/src/domain/competition/solution/fee.rs
Comment (sunce86) [reply crates/driver/src/domain/competition/solution/fee.rs]: Added check.

File: crates/driver/src/domain/competition/solution/trade.rs
Comment (sunce86) [reply crates/driver/src/domain/competition/solution/trade.rs]: Yup. updated

File: crates/driver/src/domain/competition/solution/mod.rs
Comment (sunce86) [reply crates/driver/src/domain/competition/solution/mod.rs]: renamed

File: crates/driver/src/domain/competition/solution/trade.rs
Comment (sunce86) [reply crates/driver/src/domain/competition/solution/trade.rs]: Added.

**Comment by sunce86** [comment]
> How do you want to go about adding tests? Are we going to keep the unit tests? In any case, I'd like to see a driver or integration test with the two examples we have in notion to make sure this works without coupling the test to the concrete code factoring.

I'm not going to keep the unit tests (they just show that the calculations are correct and match with the current surplus calculation we have, for example). They will exist in the PR history so we can reproduce them again in the future if needed. 

I planned on adding e2e test when everything is done. Do you think it should be part of this PR? Considering that we send no policies at the moment, merging this PR is safe.

**Comment by sunce86** [comment]
Tried testing the changes with driver tests. Due to a dynamic nature of the protocol fee was unable to figure out how to properly mock the value. Will have to test it with e2e test.

**Review by fleupold — COMMENTED** [review]
To me things are looking good, just want to see tests before merging this. Also @squadgazzz could you give this PR a review, please?

> Due to a dynamic nature of the protocol fee was unable to figure out how to properly mock the value. Will have to test it with e2e test.

What do you mean by "dynamic nature"? Doesn't the test allow you to set up specific orders, pools (and thus likely also specific fee policies)?

I'm fine with e2e tests as well, but I think they should be part of this PR (or part of the PR stack before we merge). Otherwise we tend to forget writing those tests because other things come up.

**Comment by sunce86** [comment]
> What do you mean by "dynamic nature"? Doesn't the test allow you to set up specific orders, pools (and thus likely also specific fee policies)?

I mean it's calculated based on the inputs, it's not a fixed value mocked (similar to surplus fee) which would easy the expectation asserts.

**Comment by fleupold** [comment]
> I mean it's calculated based on the inputs, it's not a fixed value mocked (similar to surplus fee) which would easy the expectation asserts.

The driver tests allow you to mock the solutions that get sent from the solver back to the driver and control the instance files that gets sent to the driver. Lastly, the `SettleOk` type let's you assert things on the final response of the driver, so you should be able to

1. Send an instance with an order (buy, sell) and a fee policy
2. Mock the solution (without protocol fee taken) from the solver to the driver
3. Inspect the final driver response and assert that the executed amounts have been adjusted accordingly

File: crates/driver/src/domain/competition/solution/mod.rs
Comment (sunce86) [reply crates/driver/src/domain/competition/solution/mod.rs:71]: I need this to coerce the error type from `fee::Error` to `SolutionError`

File: crates/driver/src/domain/competition/solution/fee.rs
Comment (sunce86) [reply crates/driver/src/domain/competition/solution/fee.rs]: added.

File: crates/driver/src/domain/competition/solution/fee.rs
Comment (sunce86) [reply crates/driver/src/domain/competition/solution/fee.rs]: removed.

**Review by fleupold — COMMENTED** [review]
🙈 I don't think the current test setup lends itself nicely for this new feature and also needs some refactoring. My goal is that we can put the tests we have in Notion in code (and not come up with completely new examples that way fewer people have thought through and agreed upon)

File: crates/driver/src/tests/cases/protocol_fees.rs
Comment (sunce86) [reply crates/driver/src/tests/cases/protocol_fees.rs]: removed tests

**Comment by sunce86** [comment]
E2E tests are extracted into a separate PR: 
https://github.com/cowprotocol/services/pull/2260

File: crates/driver/src/domain/competition/solution/fee.rs
Comment (MartinquaXD) [inline crates/driver/src/domain/competition/solution/fee.rs:136]: The smallest volume cap we can express with this is `0.5%`. Anything below that will be rounded down to 0.
Is that enough for our purposes?

File: crates/driver/src/domain/competition/solution/fee.rs
Comment (MartinquaXD) [inline crates/driver/src/domain/competition/solution/fee.rs:132]: If this is used for the surplus based fee we need to reference the quoted price somehow.
Otherwise we'll only compare against the signed limit price which already has the slippage baked in. This is not what we want, right?

File: crates/driver/src/domain/competition/solution/fee.rs
Comment (MartinquaXD) [inline crates/driver/src/domain/competition/solution/fee.rs:146]: By adding this here `executed_sell_amount_with_fee` will be greater than `sell_amount` which means `limit_buy_amount` will be greater than `executed_buy_amount` which means `surplus` will always be negative.
Unless I got it twisted.

File: crates/driver/src/domain/competition/solution/fee.rs
Comment (MartinquaXD) [inline crates/driver/src/domain/competition/solution/fee.rs:167]: nit: Might be slightly nicer to read if you only return `surplus_in_sell_token` from both `match` arms and then apply the `factor` at the end of the function because this is identical for both cases.
Same comment applies to `volume_fee()`.

File: crates/driver/src/domain/competition/solution/mod.rs
Comment (MartinquaXD) [inline crates/driver/src/domain/competition/solution/mod.rs:182]: nit: To communicate to the reader that we know upfront how many trades to expect.
I assume it was done this way because propagating errors out of `.map()` is a bit annoying.
```suggestion
        let mut trades = Vec::with_capacity(self.trades.len());
```

File: crates/driver/src/domain/competition/solution/fee.rs
Comment (sunce86) [reply crates/driver/src/domain/competition/solution/fee.rs:132]: I see the `PriceImprovement` as a general term - improvement over SOME price, whether it's a limit price or some other artificial price (artificial because `driver` works with limit orders technically and doesn't care whether it has slippage baked in and in which amount).
If it were called `MarketPriceImprovement` then it would be misleading IMO. Having a general term is better since we will implement the improvement over quotes 99.99% so renaming it would lead to renaming the autopilot/driver interface which is not too problematic right now, but it could be in the future.
OTOH, I might be biased since I spent a lot of time thinking in those terms, so if you both prefer to have `SurplusFee` I will update it.

File: crates/driver/src/domain/competition/solution/fee.rs
Comment (MartinquaXD) [reply crates/driver/src/domain/competition/solution/fee.rs:146]: Did a test computation and it looked okay.
I think I overlooked that even if you want to fill a `sell` order completely you would not set `executed == 100%` as a solver **AND** set a `surplus_fee` because then you'd spend more `sell_token` than what the user signed.
If the solver sets `executed = 100% - surplus_fee` and set `surplus_fee = some_amount` like they need to things work out.

**Review by MartinquaXD — APPROVED** [review]
Only nits remaining for me.
Approving since both sides of the renaming discussion make sense to me and I don't have strong feelings about it.
Just wanted to point it out in case this was not intentional since I remembered the plan differently. But if this is just an intermediate PR until we have all the data needed for actually taking fees over price improvement compared to quoted amounts I'm fine with it.

File: crates/driver/src/domain/competition/solution/fee.rs
Comment (sunce86) [reply crates/driver/src/domain/competition/solution/fee.rs:146]: This is enforced in the Fulfillment constructor.

File: crates/driver/src/domain/competition/solution/fee.rs
Comment (sunce86) [reply crates/driver/src/domain/competition/solution/fee.rs]: Added.

File: crates/driver/src/domain/competition/solution/fee.rs
Comment (sunce86) [reply crates/driver/src/domain/competition/solution/fee.rs]: Adjusted for 0.01%

File: crates/driver/src/domain/competition/solution/fee.rs
Comment (sunce86) [reply crates/driver/src/domain/competition/solution/fee.rs]: extracted into separate function. also refactored the big match into smaller parts - focus more on the values that are calculated.

File: crates/driver/src/domain/competition/solution/mod.rs
Comment (sunce86) [reply crates/driver/src/domain/competition/solution/mod.rs]: adjusted.

File: crates/driver/src/infra/solver/dto/solution.rs
Comment (sunce86) [reply crates/driver/src/infra/solver/dto/solution.rs]: added.

File: crates/driver/src/domain/competition/solution/trade.rs
Comment (sunce86) [reply crates/driver/src/domain/competition/solution/trade.rs]: renamed.

File: crates/driver/src/domain/competition/solution/fee.rs
Comment (sunce86) [reply crates/driver/src/domain/competition/solution/fee.rs]: I've captured the potential renaming https://github.com/cowprotocol/services/issues/2266
I believe this is ok since chances are higher that we will implement this so it's easier to close the issue than to revert the changes.

**Review by fleupold — APPROVED** [review]
Can merge this PR as is, but let's track renaming the policy as a follow up issue we should decide on and if we want to do it execute before launch.

**Pull request merged**
